### DISCUSSION TITLE
Showing legend instead of scalar bar

### DISCUSSION THREAD

**Discussion by fdinhSG** [pyvista/pyvista#4456] [Q&A] [answered]
I have a mesh where the scalar values are a discrete number, around 4 or 5 elements.

I would like to show a legend of each value instead of a continuous scalar bar. Even better, I'd like to associate a name for every element. For example, for the following sphere:

```python
import pyvista as pv
import numpy as np
mesh = pv.Sphere()
mesh['data'] = np.round((mesh.points[:,2]+.5)*3)
pl = pv.Plotter()
_ = pl.add_mesh(mesh, cmap='coolwarm')
pl.show()
```

I'd like to have a legend stating something like
- [blue] component A
- [light blue] component B
- [light red] component C
- [red] component D

And I could associate ech value with a name outside of the plotting, with a dictionary or something similar.

Is this achieveable somehow, or do I have to divide my mesh into 4 different meshes and plot them one at the time?

Thanks!

**Comment by tkoyama010** [comment, marked as answer]
You can show a legend of each value using the following code.
See more details [here](https://docs.pyvista.org/version/stable/examples/02-plot/cmap.html).

```python
import pyvista as pv
import numpy as np
from matplotlib.colors import ListedColormap

mesh = pv.Sphere()
mesh['data'] = np.round((mesh.points[:, 2] + 0.5) * 3)

mapping = np.linspace(mesh['data'].min(), mesh['data'].max(), 256)
newcolors = np.empty((256, 4))
newcolors[mapping >= 0] = np.array([0.0 / 256.0, 0.0 / 256.0, 255.0 / 256.0, 1.0])  # Blue
newcolors[mapping >= 1] = np.array([173.0 / 256.0, 214.0 / 256.0, 228.0 / 256.0, 1.0])  # Light Blue
newcolors[mapping >= 2] = np.array([255.0 / 256.0, 114.0 / 256.0, 118.0 / 256.0, 1.0])  # Light Red
newcolors[mapping >= 3] = np.array([255.0 / 256.0, 0.0 / 256.0, 0.0 / 256.0, 1.0])  # Red
colormap = ListedColormap(newcolors)

pl = pv.Plotter()
_ = pl.add_mesh(mesh, cmap=colormap)
pl.show()
```

![test](https://github.com/pyvista/pyvista/assets/7513610/6580f36a-4b20-416e-81c1-6f261652c4de)

**Reply by fdinhSG** [reply]
Thank you!

**Comment by banesullivan** [comment]
See also https://docs.pyvista.org/version/stable/api/plotting/_autosummary/pyvista.Plotter.add_legend.html#pyvista.Plotter.add_legend

**Reply by fdinhSG** [reply]
Thanks!
